src/reader.cpp

```diff
@@ -1346,86 +1363,180 @@ void reader_data_t::pager_selection_changed() {
-    const ast::string_t *matching_cmd_node = nullptr;
-    for (const node_t &n : ast) {
-        const auto *stmt = n.try_as<decorated_statement_t>();
-        if (!stmt) continue;
-
-        // Skip if we have a decoration.
-        if (stmt->opt_decoration) continue;
-
-        // See if the command's source range range contains our cursor, including at the end.
-        auto msource = stmt->command.try_source_range();
-        if (!msource) continue;
-
-        // Now see if its source range contains our cursor, including at the end.
-        if (subcmd_cursor_pos >= msource->start &&
-            subcmd_cursor_pos <= msource->start + msource->length) {
-            // Success!
-            matching_cmd_node = &stmt->command;
-            break;
+
+    parse_tree_flags_t ast_flags = parse_flag_continue_after_error |
+                                   parse_flag_accept_incomplete_tokens |
+                                   parse_flag_leave_unterminated;
+    auto ast = ast::ast_t::parse(str, ast_flags);
+
+    // Helper to check if a node is the command portion of an undecorated statement.
+    auto is_command = [&](const node_t *node) {
+        for (const node_t *cursor = node; cursor; cursor = cursor->parent) {
+            if (const auto *stmt = cursor->try_as<decorated_statement_t>()) {
+                if (!stmt->opt_decoration && node == &stmt->command) {
+                    return true;
+                }
+            }
         }
-    }
+        return false;
+    };
 
-    // Now if we found a command node, expand it.
-    maybe_t<edit_t> result{};
-    if (matching_cmd_node) {
-        assert(!matching_cmd_node->unsourced && "Should not be unsourced");
-        const wcstring token = matching_cmd_node->source(subcmd);
-        if (auto abbreviation = expand_abbreviation(token, vars)) {
-            // There was an abbreviation! Replace the token in the full command. Maintain the
-            // relative position of the cursor.
-            source_range_t r = matching_cmd_node->source_range();
-            result = edit_t(subcmd_offset + r.start, r.length, std::move(*abbreviation));
+    wcstring cmdsub_contents;
+    std::vector<positioned_token_t> result;
+    traversal_t tv = ast.walk();
+    while (const node_t *node = tv.next()) {
+        // We are only interested in leaf nodes with source.
+        if (node->category != category_t::leaf) continue;
+        source_range_t r = node->source_range();
+        if (r.length == 0) continue;
+
+        // If we have command subs, then we don't include this token; instead we recurse.
+        bool has_cmd_subs = false;
+        size_t cmdsub_cursor = r.start, cmdsub_start = 0, cmdsub_end = 0;
+        while (parse_util_locate_cmdsubst_range(str, &cmdsub_cursor, &cmdsub_contents,
+                                                &cmdsub_start, &cmdsub_end,
+                                                true /* accept incomplete */)) {
```

Comment:
This needs to check for the return value being > 0 - if it errors it returns a negative value.

That causes an infinite loop for `echo )`.